src/System.Private.CoreLib/src/System/Reflection/Emit/ILGenerator.cs

```diff
@@ -167,10 +170,10 @@ internal void InternalEmit(OpCode opcode)
             }
 
             m_ILStream[m_length++] = (byte)opcode.Value;
-
             UpdateStackSize(opcode, opcode.StackChange());
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal void UpdateStackSize(OpCode opcode, int stackchange)
```

Comment:
Do you really want to inline this method? It's huge...

Reply:
When benchmarking it made quite a significant difference, if you like I can measure the difference and let you make the call whether it is worth to keep it? I trust your instincts a lot more than mine!

Reply:
In my testing the difference is almost 2:1. Here is my [BenchmarkDotNet output](https://cdn.discordapp.com/attachments/406551726896709632/565212509670473728/bench.png)

Reply:
Will close this for now, as I feel like the performance difference is too extreme to give up. Unless someone has any objections.

Reply:
I agree that this method looks big. You should also measure the code size that each forcelined copy of this method will contribute and look into reducing it.

Reply:
I have addressed the code bloat concerns here: https://github.com/dotnet/coreclr/pull/23833/commits/c7b3b52d8e26e68d70333099744353afeb20da74.

By going with BinaryPrimitives in InternalEmit over bit-shifting approach I was able to cut down code bloat. I also factored out the cold-path of UpdateStackSize to further reduce bloat.

Reply:
I am a bit confused as to why inlining UpdateStackSize makes such a huge difference, but in every benchmark scenario I end up with massive performance regressions when remove inlining.
Does anyone know why?

|          Method | Baseline  | No Inline |   Inline  |
|---------------- |:---------:|:---------:|:---------:|
|            Emit | 33.35 ns  | 29.05 ns  | 15.03 ns  |
|   Emit_Arg_Byte | 49.03 ns  | 34.81 ns  | 21.32 ns  |
|  Emit_Arg_Short | 41.19 ns  | 36.42 ns  | 22.79 ns  |
|    Emit_Arg_Int | 51.51 ns  | 42.90 ns  | 29.41 ns  |
|   Emit_Arg_Long | 72.76 ns  | 54.20 ns  | 40.73 ns  |
| Emit_Arg_Double | 72.52 ns  | 55.55 ns  | 41.52 ns  |

Reply:
> Emit_Arg_Byte | 49.03 ns
> Emit_Arg_Short | 41.19 ns

This number looks suspect. Emit_Arg_Byte does less work than Emit_Arg_Short. Why does it take longer?

Reply:
> Does anyone know why?

I have looked at the generated code. The extra time is spent in passing the arguments around, especially `struct OpCode`.

Given that `UpdateStackSize` is internal and it is called from limited number of places, I think AggressiveInlining is reasonable here. In worst case, the total code bloat from this use AggressiveInlining is about 300 bytes that sounds acceptable.

Reply:
The naming is deceptive, Emit_Arg_Byte was actually emitting SByte. I adjusted my benchmark to test for both and the results are more in-line with what one would expect;
![Results](https://i.imgur.com/O4apsYm.png)